### PR TITLE
465 imputation links

### PR DESCRIPTION
**Pull request by KateyMatthews**
# Summary
Merging counts onto main data and pivoting by link_type.
Summing imputation_value by period, cell and question.

Add your summary here - keep it brief, to the point, and in plain English.

# Checklists

This pull request meets the following requirements:

- [ ] installable with all dependencies recorded
- [ ] runs without error
- [ ] follows PEP8 and project specific conventions
- [ ] appropriate use of comments, for example no descriptive comments
- [ ] functions documented using Numpy style docstings
- [ ] assumptions and decisions log considered and updated if appropriate
- [ ] unit tests have been updated to cover essential functionality for a reasonable range of inputs and conditions
- [ ] other forms of testing such as end-to-end and user-interface testing have been considered and updated as required
- [ ] tests suite passes (locally as a minimum)
- [ ] peer reviewed with review recorded

If you feel some of these conditions do not apply for this pull request, please
add a comment to explain why.

### CODE REVIEW

File: mbs_results/pivot_imputation_value.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
+import pandas as pd
+
+
+def merge_counts(
```
Comment (AntonZogk) [inline mbs_results/pivot_imputation_value.py:4]: I am quite sure counts are included in the post estimate dataframe

File: mbs_results/pivot_imputation_value.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    -------
+    dataframe filtered by date, containing imputation_value, pivoted by imputation type
+    and grouped by sic, cell, question and imputation type.
+
+    """
+    if selected_periods is not None:
+        df = df.query("{} in {}".format(date, selected_periods))
+
+    links_df = df.melt(
+        id_vars=[date, sic, cell, question, imputed_value],
```
Comment (AntonZogk) [inline mbs_results/pivot_imputation_value.py:113]: why imputed_value needs to be in index ?

File: mbs_results/pivot_imputation_value.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        count_df,
+        how="left",
+        left_on=[input_cell, input_date],
+        right_on=[count_cell, count_date],
+    ).astype({identifier: "int"})
+
+    return df_merge.drop(columns=[count_cell, count_date])
+
+
+def pivot_imputation_value(
```
Comment (AntonZogk) [inline mbs_results/pivot_imputation_value.py:50]: I disagree with this function name, pivot  implies reshaping data from longform to wide form, while indeed the function does that but what it returns is counts and links by sic and cell question no.

File: mbs_results/pivot_imputation_value.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        list containing periods to include in output
+
+    Returns
+    -------
+    dataframe filtered by date, containing imputation_value, pivoted by imputation type
+    and grouped by sic, cell, question and imputation type.
+
+    """
+    if selected_periods is not None:
+        df = df.query("{} in {}".format(date, selected_periods))
```
Comment (AntonZogk) [inline mbs_results/pivot_imputation_value.py:110]: Loved line 110 beautiful approach !

File: mbs_results/pivot_imputation_value.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        left_on=[input_cell, input_date],
+        right_on=[count_cell, count_date],
+    ).astype({identifier: "int"})
+
+    return df_merge.drop(columns=[count_cell, count_date])
+
+
+def pivot_imputation_value(
+    df: pd.DataFrame,
+    identifier: str,
```
Comment (AntonZogk) [inline mbs_results/pivot_imputation_value.py:52]: is identifier arg needed ?

File: mbs_results/pivot_imputation_value.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+    merged_df.drop_duplicates(inplace=True)
+    merged_df.drop(["link_type_count"], axis=1, inplace=True)
+
+    merged_df = merged_df.groupby(
+        [date, sic, cell, question, "link_type"], as_index=False
+    ).agg({imputed_value: "sum", "count": "first", "imputation_link": "first"})
+
+    sorting_order = {"F": 1, "B": 2, "C": 3}
+    merged_df["sort_column"] = merged_df["link_type"].map(sorting_order)
```
Comment (AntonZogk) [inline mbs_results/pivot_imputation_value.py:148]: Is sorting a requirement ? Nice solution, since link_type order matters we could convert that column to categorical and specify the order https://pandas.pydata.org/pandas-docs/version/1.1/reference/api/pandas.Categorical.html

File: mbs_results/pivot_imputation_value.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+    merged_df = pd.merge(
+        links_df,
+        counts_df,
+        how="outer",
+        left_on=[date, sic, cell, question, "link_type"],
+        right_on=[date, sic, cell, question, "link_type_count"],
+    )
+
+    merged_df.drop_duplicates(inplace=True)
```
Comment (AntonZogk) [inline mbs_results/pivot_imputation_value.py:140]: Can we please add a comment why this is needed ?

**Review by AntonZogk — CHANGES_REQUESTED** [review]
Nice work, all looking good.  I think we should change the function name and add some information in the docstrings too, after this happy to be merged. 

Melting and mapping is used twice , it would be nice to have a function for this.

File: mbs_results/pivot_imputation_value.py (hunk 1 above)
Comment (KateyMatthews) [reply mbs_results/pivot_imputation_value.py:4]: Thank you - yes we now realise they are. I have removed the merge_counts function and relevant input data.

File: mbs_results/pivot_imputation_value.py (hunk 3 above)
Comment (KateyMatthews) [reply mbs_results/pivot_imputation_value.py:50]: Thanks for this - agreed and name of function and relevant test data changed to 'aggregate_by_groups'.

File: mbs_results/pivot_imputation_value.py (hunk 4 above)
Comment (KateyMatthews) [reply mbs_results/pivot_imputation_value.py]: Luke's work 🙂

File: mbs_results/pivot_imputation_value.py (hunk 7 above)
Comment (KateyMatthews) [reply mbs_results/pivot_imputation_value.py:119]: We've included the drop_duplicates command to prevent incorrect summed values - please let us know if you think another approach would be better.

File: mbs_results/pivot_imputation_value.py (hunk 2 above)
Comment (KateyMatthews) [reply mbs_results/pivot_imputation_value.py]: Am I right in thinking we are going to keep this in (following discussion on Teams)?

File: mbs_results/pivot_imputation_value.py (hunk 5 above)
Comment (KateyMatthews) [reply mbs_results/pivot_imputation_value.py:52]: This has been removed; also removed references to 'identifier' in the test file.

File: mbs_results/pivot_imputation_value.py (hunk 6 above)
Comment (KateyMatthews) [reply mbs_results/pivot_imputation_value.py:127]: Do you mean something like:

merged_df['link_type'] = pd.Categorical(merged_df['link_type'], categories=['F','B','C'], ordered=True)
merged_df.sort_values(by=['link_type'], axis=0, ascending=True, inplace=True)

I'm struggling to get this (or similar) to work and pass Pytest but I could just be missing something obvious. The ordering of this column was the hardest part to get right in the first place - every time I made a change to the code it seemed to randomly change the order of link_type!

I've reverted back to the original for the time being but happy to discuss.

**Review by AntonZogk — APPROVED** [review]
All looking good, thanks for applying the changes :)
